hassio/src/components/hassio-upload-snapshot.ts

```diff
@@ -0,0 +1,174 @@
+import "@material/mwc-icon-button/mwc-icon-button";
+import { mdiFolderUpload } from "@mdi/js";
+import "@polymer/iron-input/iron-input";
+import "@polymer/paper-input/paper-input-container";
+import {
+  css,
+  customElement,
+  html,
+  internalProperty,
+  LitElement,
+  PropertyValues,
+  TemplateResult,
+} from "lit-element";
+import { classMap } from "lit-html/directives/class-map";
+import { fireEvent } from "../../../src/common/dom/fire_event";
+import "../../../src/components/ha-circular-progress";
+import "../../../src/components/ha-svg-icon";
+import { extractApiErrorMessage } from "../../../src/data/hassio/common";
+import { uploadSnapshot } from "../../../src/data/hassio/snapshot";
+import { HomeAssistant } from "../../../src/types";
+
+declare global {
+  interface HASSDomEvents {
+    "snapshot-uploaded": undefined;
+  }
+}
+
+@customElement("hassio-upload-snapshot")
+export class HassioUploadSnapshot extends LitElement {
+  public hass!: HomeAssistant;
+
+  @internalProperty() public value: string | null = null;
+
+  @internalProperty() private _error = "";
+
+  @internalProperty() private _uploading = false;
+
+  @internalProperty() private _drag = false;
+
+  protected updated(changedProperties: PropertyValues) {
+    if (changedProperties.has("_drag") && !this._uploading) {
+      (this.shadowRoot!.querySelector(
+        "paper-input-container"
+      ) as any)._setFocused(this._drag);
+    }
+  }
+
+  public render(): TemplateResult {
+    return html`
+      ${this._uploading
+        ? html`<ha-circular-progress
+            alt="Uploading"
+            size="large"
+            active
+          ></ha-circular-progress>`
+        : html`
+            ${this._error ? html`<div class="error">${this._error}</div>` : ""}
+            <label for="input">
+              <paper-input-container
+                .alwaysFloatLabel=${Boolean(this.value)}
+                @drop=${this._handleDrop}
+                @dragenter=${this._handleDragStart}
+                @dragover=${this._handleDragStart}
+                @dragleave=${this._handleDragEnd}
+                @dragend=${this._handleDragEnd}
+                class=${classMap({
+                  dragged: this._drag,
+                })}
+              >
+                <label for="input" slot="label">
+                  Upload snapshot
+                </label>
+                <iron-input slot="input">
+                  <input
+                    id="input"
+                    type="file"
+                    class="file"
+                    accept="application/x-tar"
+                    @change=${this._handleFilePicked}
+                  />
+                </iron-input>
```

Comment:
Can we take the guts of the file uploader out of here in a separate element and use it in both picture upload and this uploader?

Reply:
Like all the handledrop, handle drag and the paper input elements

Reply:
Sure :+1: